BugFix and Evol MouseEvents

jefleponot opened the pull request:
Hi @istr , @mickaelandrieu , others

When we use Native Events something doesn't work. In fact we can't use these events on selector when there aren't in the visible area (inside the page viewport and scrollPosition areas). It doesn't work also if scroll position is not fit page correctly. So it is not user-friendly ... because you can't maximize viewport without having some decrease in reactivity (SlimerJs).

"ContextMenu" doesn't work except onto the main html element.

With this PR you can control more finely event on canvas, image ...

jefleponot commented:
Travis command is out ?

istr commented:
@jefleponot You introduced some linting errors. See commit comments.

istr commented on modules/mouse.js:
```diff
@@ -136,6 +148,51 @@ var Mouse = function Mouse(casper) {
+                    var struct = casper.page.evaluate(function (clientX, clientY) {
+						function xpath(el) {
+                            if (typeof el == "string") return document.evaluate(el, document, null, 0, null);
+                            if (!el || el.nodeType != 1) return '';
```
Lint error. Use ```!==``` instead of ```!=```

mickaelandrieu commented on modules/clientutils.js:
```diff
@@ -661,25 +696,37 @@
                 this.log("mouseEvent(): Couldn't find any element matching '" + selector + "' selector", "error");
                 return false;
             }
+            // 
```
to be removed

istr commented on modules/mouse.js:
```diff
@@ -136,6 +148,51 @@ var Mouse = function Mouse(casper) {
+                    }
+                    var struct = casper.page.evaluate(function (clientX, clientY) {
+						function xpath(el) {
+                            if (typeof el == "string") return document.evaluate(el, document, null, 0, null);
```
Lint error. Use ```===``` instead of ```==```.

mickaelandrieu commented on docs/modules/casper.rst:
```diff
@@ -1357,16 +1368,16 @@ Retrieves current page title::
 
 Triggers a mouse event on the first element found matching the provided selector.
 
-Supported events are ``mouseup``, ``mousedown``, ``click``, ``mousemove``, ``mouseover`` and ``mouseout``::
```
imo you should add a new "versionadded" section, and talk about phantomjs <= 1.9.8 limitations

istr commented on modules/mouse.js:
```diff
@@ -136,6 +148,51 @@ var Mouse = function Mouse(casper) {
+                            if (typeof el == "string") return document.evaluate(el, document, null, 0, null);
+                            if (!el || el.nodeType != 1) return '';
+                            if (el.id) return "//*[@id='" + el.id + "']";
+                            var sames = [].filter.call(el.parentNode.children, function (x) { return x.tagName == el.tagName; });
```
Lint error. Use ```===``` instead of ```==```.

mickaelandrieu commented on tests/site/area.html:
```diff
@@ -0,0 +1,99 @@
+<script src="http://www.kirupa.com/prefixfree.min.js"></script>
+<script>
+	(function(window) {
+var theThing = document.querySelector("#thing");
```
indentation issue here

jefleponot commented:
thanks you. I do it right now :-)

jefleponot commented:
copy paste in not always a good thing :eyes:

mickaelandrieu commented:
> copy paste in not always a good thing :eyes:

copy/paste is never a good thing ;)

istr commented on modules/casper.js:
```diff
@@ -458,13 +458,16 @@ Casper.prototype.clear = function clear() {
  *
  * @param  String   selector  A DOM CSS3 compatible selector
+ * @param  String   target    A HTML target '_blank','_self','_parent','_top','framename' (optional)
+ * @param  {Number} x         X position (optional)
```
This documentation does not match the function signature.

istr commented on modules/clientutils.js:
```diff
@@ -242,14 +245,15 @@
-                errors: [],
-                fields: [],
-                files:  []
+                "errors": [],
```
Don't do this. Please fix the linting rule instead if this produces a warning.

jefleponot replied:
is it ok ?

istr replied:
No, please set
```"quote-props": ["as-needed", { "keywords": true }]``` in ```.eslintrc``` instead. This has the effect to reduce the change-set in this PR and probably the nice side-effect to get rid of lots of lint warnings in one go.

istr replied:
revert

istr commented:
Ok, wrt the comments in #1413 I totally agree with @jefleponot. This one is a real bug-fix I would like to merge after some cleanup; see code comments.

mickaelandrieu commented on docs/modules/casper.rst:
```diff
@@ -1359,16 +1370,21 @@ Retrieves current page title::
+**Signature:** ``mouseEvent(String type, String selector, [Number|String X, Number|String Y])``
 
 .. versionadded:: 0.6.9
+.. versionupdate:: 1.1.6
```
1.1.0, we are always in beta state, no ?

istr commented on docs/modules/casper.rst:
```diff
@@ -1359,16 +1370,21 @@ Retrieves current page title::
+and for phantomjs >= 1.9.8 ``mouseenter``, ``mouseleave`` and ``contextmenu``::
+
+.. warning::
+Events depend on version Browser - to generate all kinds of Popular Events use the latest phantomJS or SlimerJS versions
```
"The list of supported events depends on the version of the engine in use. Older engines only provide partial support. For best support use recent builds of PhantomJS or SlimerJS."

istr commented on docs/modules/clientutils.rst:
```diff
@@ -135,7 +135,26 @@ Retrieves a single DOM element by a :ref:`selector expression <selectors>`::
+**Signature:** ``forceTarget(String selector, String target)``
+
+Force a browsing to use another target instead the one provided. Very useful to limitate window number and reduce memory::
+
```
"Force the engine to use another target instead of the one provided. Very useful to limit the number of open windows and reduce memory consumption."

istr commented on docs/modules/clientutils.rst:
```diff
@@ -135,7 +135,26 @@ Retrieves a single DOM element by a :ref:`selector expression <selectors>`::
+
+Force a browsing to use another target instead the one provided. Very useful to limitate window number and reduce memory::
+
+    var href;
```
drop this line

istr replied:
Is this global variable necessary for the example? Could you make it local to the step() function?

istr commented on docs/modules/clientutils.rst:
```diff
@@ -195,6 +214,26 @@ Retrieves current document height::
+
+Retrieves current document width::
+
+    var documentHeight;
```
Where is this used?

jefleponot replied:
in mouse.js

istr commented on modules/casper.js:
```diff
@@ -1206,7 +1209,8 @@ Casper.prototype.handleReceivedResource = function(resource) {
         return;
     }
     this.resources.push(resource);
-    if (utils.decodeUrl(resource.url) !== this.requestUrl) {
```
Please check if this still works with PhantomJS 2.1.0, see #1495.

istr commented on modules/clientutils.js:
```diff
@@ -364,6 +370,25 @@
         };
 
+         /**
+         * Force target on <FORM> and <A> tag.
```
one step left

istr commented on modules/mouse.js:
```diff
@@ -45,23 +40,65 @@ var Mouse = function Mouse(casper) {
-            var x = Math.round(bounds.left + bounds.width / 2);
-            var y = Math.round(bounds.top  + bounds.height / 2);
+            var x = Math.floor(bounds.left + bounds.width / 2),
+                y = Math.floor(bounds.top + bounds.height / 2);
```
Please stick with one variable. What is the reason to change it to ```floor``` here?

istr commented:
Great! Apart from some very minor tweaks to the docs and some tiny questions I am ok to merge this. See code comments.

istr commented:
@jefleponot Would you mind addressing the eslint issue I mentioned? I would like to merge this.

jefleponot commented:
Hi @istr , I merged with the eslint issue you mentioned but it doesn't works

/usr/local/lib/node_modules/eslint/lib/config/config-validator.js:102
        throw new Error(message.join(""));
              ^
Error: /casperjs/.eslintrc:
	Configuration for rule "quote-props" is invalid:
	Value "as-needed,[object Object]" no schemas match

istr commented:
You might need to
- update eslint (```npm -g update eslint```)
- use the following rule ```"quote-props": [1, "as-needed", {"keywords": true}],```

The rule did not work for me with eslint 1.7.* and showed your error message, but works with
```
$ eslint --version
v2.7.0
```

jefleponot commented:
Perfect! Everything works! Good job @istr! :+1:

istr commented on modules/clientutils.js:
```diff
@@ -242,14 +245,15 @@
-                fields: [],
-                files:  []
+                "errors": [],
+                "fields": [],
```
revert

istr commented on modules/clientutils.js:
```diff
@@ -242,14 +245,15 @@
-                files:  []
+                "errors": [],
+                "fields": [],
+                "files": []
```
revert

istr commented on modules/clientutils.js:
```diff
@@ -270,20 +274,21 @@
 
             var finders = {
-                css: function(inputSelector, formSelector) {
+                "css": function(inputSelector) {
```
revert quoting

istr commented on modules/clientutils.js:
```diff
@@ -270,20 +274,21 @@
-                labels: function(labelText, formSelector, value) {
-                    var label = this.findOne({type: "xpath", path: '//label[text()="' + labelText + '"]'}, form);
-                    if(label && label.htmlFor) {
+                "labels": function(labelText) {
```
revert quoting

istr commented on modules/clientutils.js:
```diff
@@ -270,20 +274,21 @@
                     }
                 },
-                names: function(elementName, formSelector) {
+                "names": function(elementName) {
```
revert quoting

istr commented on modules/clientutils.js:
```diff
@@ -270,20 +274,21 @@
                 },
-                xpath: function(xpath, formSelector) {
-                    return this.findAll({type: "xpath", path: xpath}, form);
+                "xpath": function(xpath) {
```
revert quoting

istr commented on modules/clientutils.js:
```diff
@@ -294,23 +299,24 @@
-                            type: findType,
-                            selector: selector,
-                            path: err.path
+                            "type": findType,
```
revert quoting this and following two lines

istr commented on modules/clientutils.js:
```diff
@@ -389,11 +414,12 @@
             try {
                 return this.sendAJAX(url, method, data, false, {
-                    overrideMimeType: "text/plain; charset=x-user-defined"
+                    "overrideMimeType": "text/plain; charset=x-user-defined"
```
revert

istr commented on modules/clientutils.js:
```diff
@@ -427,10 +467,10 @@
-                    left:   clipRect.left,
-                    width:  clipRect.width,
-                    height: clipRect.height
+                    "top": clipRect.top,
```
revert quoting this and following three lines

istr commented on modules/clientutils.js:
```diff
@@ -450,15 +490,14 @@
-                        left:   clipRect.left,
-                        width:  clipRect.width,
-                        height: clipRect.height
+                        "top": clipRect.top,
```
revert quoting this and following three lines

istr commented on modules/clientutils.js:
```diff
@@ -480,16 +519,16 @@
-                width: bounds.width,
-                height: bounds.height,
-                visible: this.visible(selector)
+                "nodeName": element.nodeName.toLowerCase(),
```
revert quoting this and following nine lines

istr commented on modules/clientutils.js:
```diff
@@ -508,16 +547,16 @@
-                    width: bounds[index].width,
-                    height: bounds[index].height,
-                    visible: eleVisible(element)
+                    "nodeName": element.nodeName.toLowerCase(),
```
revert quoting this and following nine lines

istr commented on modules/clientutils.js:
```diff
@@ -857,15 +914,15 @@
-                                name:    "FileUploadError",
-                                message: "File field must be filled using page.uploadFile",
-                                path:    value
+                                "name": "FileUploadError",
```
revert quoting this and following two lines

istr commented:
Great. I marked (hopefully all) lines that need to be reverted.

jefleponot commented:
/home/jef/dev/casperjs/modules/clientutils.js
   33:1   warning  Rule 'no-empty-label' was removed and replaced by: no-labels                 no-empty-label
   33:1   warning  Rule 'space-return-throw-case' was removed and replaced by: keyword-spacing  space-return-throw-case
   33:1   warning  Rule 'space-after-keywords' was removed and replaced by: keyword-spacing     space-after-keywords
   75:17  warning  Expected no return value                                                     consistent-return
   78:17  warning  Expected no return value                                                     consistent-return
  280:25  warning  Expected to return a value at the end of this function                       consistent-return
  311:25  warning  Blocks are nested too deeply (4)                                             max-depth
  338:33  warning  Expected to return a value at the end of this function                       consistent-return
  359:33  warning  Expected to return a value at the end of this function                       consistent-return
  466:42  warning  Expected to return a value at the end of this function                       consistent-return
  491:43  warning  Expected to return a value at the end of this function                       consistent-return
  546:53  warning  Too many nested callbacks (2). Maximum allowed is 1                          max-nested-callbacks
  571:43  warning  Expected to return a value at the end of this function                       consistent-return
  626:74  warning  Unexpected use of undefined                                                  no-undefined
  630:46  warning  Expected to return a value at the end of this function                       consistent-return
  664:32  warning  Unexpected use of undefined                                                  no-undefined
  844:25  warning  Blocks are nested too deeply (4)                                             max-depth
  905:33  warning  Blocks are nested too deeply (4)                                             max-depth
  923:33  warning  Blocks are nested too deeply (4)                                             max-depth

istr commented:
Thank you so much, @jefleponot. I am now :+1: to merge this. @mickaelandrieu, @BIGjuevos your turn.

jefleponot commented:
You are welcome

BIGjuevos commented:
I like this.  Thanks for solving the problem @jefleponot :+1:

mickaelandrieu commented:
Merged, big thanks @jefleponot !

jefleponot commented:
my second contribution :-) Hip Hip Hip Houra
